### PR TITLE
implement post log to paste server

### PR DESCRIPTION
**Pull request by bill-auger**
this addresses feature request #1182

this commit adds a button to the failure MessageBox which posts the install log to termbin.com - this approach has the extra advantage of the back-end (fiche) being freely licensed and easy to self host, if one wants the posts to be private, for example

### CODE REVIEW

File: src/libcalamaresui/ViewManager.cpp [hunk 1]
```diff
@@ -189,28 +192,106 @@ ViewManager::insertViewStep( int before, ViewStep* step )
 void
 ViewManager::onInstallationFailed( const QString& message, const QString& details )
 {
+bool    shouldOfferWebPaste = true; // TODO: config var
```
Comment (abucodonosor) [inline src/libcalamaresui/ViewManager.cpp]: Btw at least that TODO need be implemented.

Without that we force that on all the time.

**Comment by bill-auger** [comment]
yes, this was offered for discussion - it includes some naught
bits in its current form that should not be merged

it is also not as optimal as it could be - it is not my original
implementation - i re-wrote it almost entirely yesterday
specifically for this, to avoid shelling out to netcat; so it is
not well-tested

for examples:

* i havent looked, but the log data may actually be in memory
already (or it could be made to be); so the filesystem access
could be eliminated

* it may be possible to avoid spawning the second QMessageBox*
pasteUrlMsgBox; though it was not obvious how

* with all of the error checking, the code style is violated due
to so many indentation levels; so it may be better to refactor
that lambda into a new method

* in its current form, it segfaults intermittently due to a
MUTEX issue; though i suspect that is due to forcing the method
to fire from the ViewManager constructor, at such an early stage
while the requirements check is running

**Comment by bill-auger** [comment]
yes, this was offered for discussion - it includes some naughty bits, in its current form, that should not be merged

it is also not as optimal as it could be - it is not my original implementation - i re-wrote it almost entirely yesterday specifically for this, to avoid shelling out to netcat; so it is not well-tested

examples for discussion:

* it inttroduces several new strings that need translations:
    * QString pasteMsg = tr( "Would you like to paste the install log to the web?" );
    * QString pasteUrlFmt = tr( "Install log posted to: %1" );
    * QString pasteUrlTitle = tr( "Install Log Paste URL" );
    * tr( "&Yes" ) )

* i havent looked, but the log data may actually be in memory already (or it could be made to be); so the filesystem access could be eliminated

* it may be possible to avoid spawning the second QMessageBox, for presenting the paste URL, by preventing the existing one from closing; though it was not obvious how to - that may entail overriding the 'onClicked' handler in a custom widget

* one error check was omitted - socket->readLine() returns the number of bytes read - a result of 0 or -1 indicates that nothing was read; in which case it should probably indicate the failure, rather than presenting the "Install log posted to: %1" message - also, it should return a valid URL; so that could be validated - (UPDATE: i added the data length check)

* it is not obvious if the "new QTcpSocket" and/or the "new QMessageBox" should be deleted explicitly; though the existing QMessageBox is not - perhaps some magic handles that; but of course, the program exits immediately afterward

* with all of the error checking, the code style is violated due to so many indentation levels; so it may be better to refactor that lambda into a new method (UPDATE: i just re-worked this - it is much more readable with early returns; though im not sure if that is acceptable style)

* in its current form, it segfaults intermittently due to a MUTEX issue; though i suspect that is due to forcing the method to fire from the ViewManager constructor, at such an early stage while the requirements check is running

**Comment by bill-auger** [comment]
UPDATE:

* added the socket read data length check and a crude URL validation

* unwrapped the nested error checks - it is much more readable with early returns; though im not sure if that is acceptable style)

**Comment by adriaandegroot** [comment]
Thanks, @bill-auger . This one is going to wait for the next release cycle, early august.

**Pull request merged**